examples/data_preprocessing/police_killing/preprocessing_rules_from_paper.ipynb

```diff
@@ -0,0 +1,1009 @@
+   "outputs": [],
+   "source": [
+    "\"\"\"\n",
+    "manually creating a label2label_id dict, since there are only two classes, \n",
+    "and a label_id2label dict (I will need that later, but I'm not sure if I'm allowed to to it like this)\n",
+    "Is it okay to create it manually?\n",
+    "\"\"\"\n",
+    "\n",
+    "label2label_id ={\"negative\":0, \"positive\":1}\n",
+    "label_id2label = {0: \"negative\", 1: \"positive\"}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## building the T matrix\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 40,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "num_classes = 2"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 41,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\"\"\"\n",
+    "mapping to t matrix \n",
+    "(from TAC tutorial)\n",
+    "\"\"\"\n",
+    "\n",
+    "def get_mapping_rules_labels_t(rule2label, num_classes):\n",
+    "    \"\"\" Function calculates t matrix (rules x labels) using the known correspondence of relations to decision rules \"\"\"\n",
+    "    mapping_rules_labels_t = np.zeros([len(rule2label), num_classes])\n",
+    "    for rule, labels in rule2label.items():\n",
+    "        mapping_rules_labels_t[rule, labels] = 1\n",
+    "    return mapping_rules_labels_t\n",
+    "\n",
+    "mapping_rules_labels_t = get_mapping_rules_labels_t(rule2label, num_classes)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## buidling the Z matrix (instances x rules)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**getting the train data**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 42,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def rules2regex(rule2rule_id):\n",
+    "    \n",
+    "    searches = dict({})\n",
+    "    \n",
+    "    for rule in rule2rule_id.keys():\n",
+    "            wordpair = rule.split()\n",
+    "            search4police_word = f'{wordpair[0]}\\W'\n",
+    "            search4kill_word = f'{wordpair[1]}\\W'\n",
+    "            searches[rule] = [search4police_word, search4kill_word]\n",
+    "            \n",
+    "    return searches\n",
+    "\n",
+    "\n",
+    "def get_data_for_df(data, searches):\n",
+    "\n",
+    "    data_for_df = []\n",
+    "\n",
+    "    for sample in tqdm(data[\"samples\"].drop_duplicates()):\n",
+    "        data_dict = dict({})\n",
+    "        data_dict[\"samples\"] = sample\n",
+    "        data_dict[\"rules\"] = []\n",
+    "        data_dict[\"enc_rules\"] = []\n",
+    "        \n",
+    "        #if \"label\" in data.columns:\n",
+    "            #data_dict[\"label\"] = data.loc[data.samples == sample, 'label'].values[0]\n",
+    "\n",
+    "\n",
+    "#this would have been my idea to keep the labels in the test data. But it makes it very slow. (8 instead of less than 1 mins)\n",
```

Comment:
I don't think I understand what you mean. Could you explain pls (if it is relevant)?